### PR TITLE
Pass --arch and --platform flag to the 'lifecycle' hook to support cross-compilation (or proper download).

### PR DESCRIPTION
**Pull request by sfc-gh-ptabor**
Fixes: #1225

pnpm 'install' gets proper platform flags (--arch & --platform) such that right dependencies are being built/downloaded. 

Extensive description in: #1225.

### Type of change

- Bug fix (change which fixes an issue)
- (with a small test of:) New feature or functionality (change which adds functionality)

### Test plan

Will update `e2e/js_image_oci` if the change is considered OK.

### CODE REVIEW

**Comment by sfc-gh-ptabor** [comment]
@alexeagle Hi. Could you, please take a look ?

**Comment by sfc-gh-ptabor** [comment]
@gregmagolan @alexeagle: Could you, please take, a look.

File: npm/private/npm_import.bzl
Comment (gregmagolan) [inline npm/private/npm_import.bzl:160]: Clever way to wire it up 👍

File: npm/private/lifecycle/lifecycle-hooks.js
Comment (gregmagolan) [inline npm/private/lifecycle/lifecycle-hooks.js:190]: Nice. I wasn't aware you could set these here. I tracked this down to ending up in `config` on the `lifecycle` call to `@pnpm/npm-lifecycle` and ending up the `env`. Could leave a brief comment here as to how it works?

Comment (gregmagolan) [reply npm/private/lifecycle/lifecycle-hooks.js:190]: Links to docs or directly to code would be helpful for the next person who looks at this

File: npm/private/npm_import.bzl
Comment (sfc-gh-ptabor) [reply npm/private/npm_import.bzl:160]: Thank you.

File: npm/private/lifecycle/lifecycle-hooks.js
Comment (sfc-gh-ptabor) [reply npm/private/lifecycle/lifecycle-hooks.js:190]: Added.

**Comment by sfc-gh-ptabor** [comment]
@gregmagolan Please Take Another Look.

**Comment by sfc-gh-ptabor** [comment]
@gregmagolan - are the new descriptions sufficient ?

**Review by gregmagolan — APPROVED** [review]
🚀

**Comment by gregmagolan** [comment]
> @gregmagolan - are the new descriptions sufficient ?

Those are great. Thanks for updating. Running CI and will merge when green.

**Review by gregmagolan — CHANGES_REQUESTED** [review]
Hmmm. Looks like this is going to need some more work to be compatible with bzlmod. Looks issues accessing `@platform` and with the select statements. You should be able to reproduce locally with you run with `--enable_bzlmod`.

**Comment by sfc-gh-ptabor** [comment]
I'm not sury why, but bumping the version of `platforms` to 0.0.5 resolved the problem for me locally.

**Comment by gregmagolan** [comment]
> I'm not sury why, but bumping the version of `platforms` to 0.0.5 resolved the problem for me locally.

That does ring a bell with platforms 0.0.4 having issues with bzlmod tho I can't recall the details. I've seen a problem of a different shape with platforms and bzlmod before.

I pushed a commit that added the dep on "platforms" to all of the failing e2e tests. Want to see if that is sufficient to green them up. It is an issue, however, that this change adds that dependency on users since that would make it a BREAKING CHANGE. I think we'll have to figure out how to wire it up so that "platforms" dep is internal to rules_js and the generated starlark that ends up in the user's workspace doesn't have any references to `@platforms`.

**Comment by gregmagolan** [comment]
Yeah, looks all green with the added `@platform` deps on the e2e tests. So we just have to figure out how to wire it up so that those are not needed.

I think aliasing the `@platform` targets in use in `@aspect_rules_js` and using those in the generated code instead will work.

**Comment by sfc-gh-ptabor** [comment]
Thank you @gregmagolan for the advice and review.

 I added the platforms. I did it in the pre-existing file: //platforms.
Alternative is to have it as part of `/js/private/(platforms)?/` package.

Please approve the tests to make sure it works.

**Comment by gregmagolan** [comment]
That looks great. I'm fine with them going under `//platforms`. I pushed a small tweak so the targets are under `//platforms/os/` and `//platforms/cpu/` so they match closer with the `@platforms` repository.

Looks good to me. Will let @alexeagle have a look before merging.

**Review by gregmagolan — APPROVED** [review]

**Review by alexeagle — COMMENTED** [review]
I think this works for your use case, but it's a bit naive.

This unconditionally passes platform,arch to the lifecycle rawConfig which means any package that has compilation will be affected. If a platform-dependent npm binary is run in a build action, it be built for the target platform and not work. In theory, we need three different actions:
- run the lifecycle hooks for the host platform so the tool can be used in a local action
- run the lifecycle hooks for the exec platform so the tool can be used in a remote action
- run the lifecycle hooks for the target platform so the library can be packaged for runtime

Maybe one example is the `esbuild` package which has platform-dependent `optionalDependencies`, that seems like something we should test before landing this. Maybe we'll end up with a linux esbuild binary when trying to bundle on a mac.

**Comment by gregmagolan** [comment]
Varying host/exec platform with RBE doesn't work in general with Bazel even at the toolchain layer so I don't think it is a use case we need to solve for right now. If you're bazel running a target on the host then target platform will be set to the host and it should work as expected.

I think this is an improvement over what we have now which is that setting or transitioning the target platform, for example if you're building for a docker container on a mac, then this change will allow for cross-compilation.

File: platforms/os/BUILD.bazel
Comment (thesayyn) [inline platforms/os/BUILD.bazel:1]: i don't see the value here for these aliases. why can't we just use them in the select statement?

Comment (gregmagolan) [reply platforms/os/BUILD.bazel:1]: bzlmod isn't happy about that because the select statements ends up in the user's workspace which forces the user to dep on `platforms` in their MODULE.bazel

Comment (thesayyn) [reply platforms/os/BUILD.bazel:1]: oh, i see that makes sense. in that case can we move these into private?

Comment (gregmagolan) [reply platforms/os/BUILD.bazel:1]: We could but they would then go in `npm/private` since they are being by `npm`; we may want to use them in `js/` the future so I figure `platforms` is a better home. Perhaps `platforms/private` tho we don't advertise `//platforms` as public API anywhere.

**Comment by alexeagle** [comment]
Even without remote execution taken into consideration, won't this always run lifecycle hooks for the target platform even for a package that you intend to use with `js_run_binary` in a build step? I think that works today but will fail with this change.

**Comment by gregmagolan** [comment]
> Even without remote execution taken into consideration, won't this always run lifecycle hooks for the target platform even for a package that you intend to use with `js_run_binary` in a build step? I think that works today but will fail with this change.

`js_run_binary` will transition the `js_binary` it is running to the execution platform along with all of its deps. It should work as expected with this change.

**Comment by sfc-gh-ptabor** [comment]
1. @alexeagle Thank you for taking a look. I'm likely missing some, piece, but thought that executionTransitions
(https://github.com/bazelbuild/proposals/blob/main/designs/2019-02-12-execution-transitions.md) is the answer to the problem you are signaling, i.e. target should always be build for the 'target' platform, but it's responsibility of infrastructure to transition target platform to proper execution platform if a target is going to be used as 'tool'. 

2. How about `//private/platforms/{cpu,os}` for both: 
    - signaling it's not a public API
    - mimicking @platforms as closely as possible ?

**Comment by sfc-gh-ptabor** [comment]
@alexeagle @gregmagolan @thesayyn: Could, you, please comment whether `//private/platforms/{cpu,os}` is the desired path ?

File: platforms/os/BUILD.bazel
Comment (alexeagle) [reply platforms/os/BUILD.bazel:1]: I don't think this is a meaningful "public API surface" to bother about the private distinction.

**Review by alexeagle — APPROVED** [review]

**Comment by alexeagle** [comment]
A user reports errors in some of their builds, likely related to this change:

```
configurable attribute "args" in //fullscreen:.aspect_rules_js/node_modules/source-map@0.1.43/lc doesn't match this configuration. Would a default condition help?
2023-09-27 17:39:57 UTC	
2023-09-27 17:39:57 UTC	Conditions checked:
2023-09-27 17:39:57 UTC	 @aspect_rules_js//platforms/os:osx
2023-09-27 17:39:57 UTC	 @aspect_rules_js//platforms/os:linux
2023-09-27 17:39:57 UTC	 @aspect_rules_js//platforms/os:windows
```

**Comment by jbedard** [comment]
@sfc-gh-ptabor would you be interested in fixing the error that user is reporting? I think we just need to add a default value when the os/architecture is unknown, basically falling back to the previous behaviour in that case?

**Comment by sfc-gh-ptabor** [comment]
@jbedard @alexeagle the fix is in: https://github.com/aspect-build/rules_js/pull/1297

Alternative is to learn which specific arch/os user needs and add explicit values for that mapping.

**Pull request merged**
